server/src/main/java/org/elasticsearch/rest/RestRequest.java

```diff
@@ -104,6 +106,23 @@ private RestRequest(NamedXContentRegistry xContentRegistry, Map<String, String>
         this.requestId = requestId;
     }
 
+    private static @Nullable ParsedMediaType parseHeaderWithMediaType(Map<String, List<String>> headers, String headerName) {
+        //TODO: make all usages of headers case-insensitive
```

Comment:
this is a not well documented fact, but by default Netty uses a case insensitive hashing strategy for retrieval of headers. Checkout the `DefaultHttpHeaders` class if you're interested. Due to our use of Netty for HTTP parsing we also get this benefit in the nio transport. Also, see the `Netty4HttpRequest`/`NioHttpRequest`.`HttpHeadersMap` class.

Reply:
Interesting, I did not know about this. 
I wrongly relied on tests only using map from java collections.
Maybe we might want to refactor usages of java.util.Map to an interface that documents/enforces case insensitivity? 
I will leave this todo for now so that we can reconsider this in the future